### PR TITLE
Test of final state

### LINKED ISSUE #156: DPMJET stable/unstable settings
There are couple of problems with setting particles for decay in DpmjetIII:

* If 
```python
for pdg in unstable_pids:
    dpmjet_generator.set_unstable(pdg)
```
is excecuted before any event generation, then the function `self._lib.dt_kkinc` in `_generate()` reset stable/unstable settings. It can be checked by inspecting the content of `self._lib.pydat3.mdcy[:, 0]` before and after the first call. In some cases stable/unstable settings are reset after the first call as well. It seems that reset occurs after each call for a new nucleus.

* Some particles (Sigma~+(-3112), Sigma-(3112), Sigma~-(-3222), Sigma+(3222), Xi~+(-3312), Xi-(3312), Xi~ 0(-3322), Xi0(3322), Omega~+(-3334), Omega-(3334), and some other) are still present as final particles, although `self._lib.pydat3.mdcy[:, 0]` shows that they are set as unstable. Partly, the problem is solved by setting `event_generator._lib.pydat1.mstj[22 - 1] = 1`, which sets decay for all particles that are set as unstable. For some reason, Dpmjet constructor sets `event_generator._lib.pydat1.mstj[22 - 1] = 2`. In that case only particles with proper time decay smaller than 10 mm decays. Setting long enought time by `event_generator._lib.pydat1.parj[71 - 1] = 1e10` should set all unstable particles for decay. However, in both cases Sigma-(3112),  Sigma0(3212), Sigma+(3222) still present in output. It can be solved, only if all particles are set as unstable by `self._lib.pydat3.mdcy[:, 0] = 1`, which in most cases don't want to do, as some particles should still be present in output (for example muons).  The problem seems appear only for nuclei. For proton as a target, Sigma particles disappear.

The first step to change the behaviour is to set `event_generator._lib.pydat1.mstj[22 - 1] = 1`, unless `event_generator._lib.pydat1.mstj[22 - 1] = 2` has special meaning. In that case we can add an interface function for `event_generator._lib.pydat1.parj[71 - 1]`.

The problem with not decaying Sigma particles is somewhere in Dpmjet and related somehow to nuclei.

For a record, I tried to set as unstable the particles with following pdg ids: 

```python
unstable_pids = [441, -3112, 3112, -3212, 3212, -3222,  3222, -3312, 3312, -3322, 3322, -3334, 3334, -4112, 4112, -4122, 4122, -4212, 4212, -4222, 4222]
```

### PR DESCRIPTION
**Pull request by HDembinski**
This PR fixes the issues with DPMJET and PHOJET models, which didn't follow the decay settings controlled by chromo, and adds a test which checks that all particles which we expect to occur in the final state actually occur for all models.

Two exceptions are built into the test:
- SIBYLL-2.1 does not generate Omega- (and cc)
- UrQMD doesn't generate only Omega- (however this might be related to limited stats) 
- The QGSJet family does not generate Omega-, Xi0, Xi-, Sigma+, Sigma- (and cc)

The PR fixes the following problems to make the test pass:
- [x] UrQMD does not generate KS particles. Bug was fixes in our fortran middle layer.
- [x] DPMJetIII and PHOJET were not following the decay settings requested. Bugs have been identified and fixes to DPMJET backported + submodules updated.
- [x] Vitange DPMJET/PHOJET had the problem as well, and it was fixed in `chromo`. The minor version of both codes have been upped by one to reflect the minor change.
- [x] All tests should pass with less XFAILs for decays in PHOJET/DPMEJT models

This PR closes #156.

### CODE REVIEW

**Comment by afedynitch** [comment]
So, I found the issues with PHOJET and likely with DPMJET decays, and the mystery difference between 19.1 and 19.3. I actually cleaned up these subroutines, not perfectly, but there is some improvement :)

I need to backport 19.3 to 19.1 because no Omega- is definitely a bug.

**Comment by afedynitch** [comment]
@HDembinski Sorry, I hijacked your PR to make substantial fixed to the decays in all of the PHOJET/DPMJET models. I identified the issues and fixed them directly in the fortran codes and also in chromo. I can't request a review from you, but nonetheless I think you can do a review or comment if you wish.

I also changed the base of the PR to be the branch we're preparing for release.

File: tests/test_final_state.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    msg = f"{Model.pyname} zero counts for "
+    msg += ", ".join(
+        pdg2name(pid)
+        for pid, n, ki in zip(long_lived_hadrons, counts, known_issues)
+        if n == 0 and not ki
+    )
+    test = np.all((counts > 0) ^ known_issues)
+    assert test, (
+        msg
+        + ", ".join(
```
Comment (jncots) [inline tests/test_final_state.py:71]: It can be added to msg above.

**Review by jncots — COMMENTED** [review]
Cool. Finally decaying particles will decay.

File: src/fortran/urqmd-3.4/chromo_urqmd.f [hunk 2]
```diff
@@ -64,7 +64,7 @@ SUBROUTINE CHEPEVT
 c We will replace those with K0S/L based on a 50/50 rule.
 
          IF (ABS(IDHEP(I)).EQ.311) THEN
-            IDHEP(I) = K0SEL(INT(2.D0*SIMRND()))
+            IDHEP(I) = K0SEL(1 + INT(2.D0*SIMRND()))
```
[the remark this answers is not in the document]
Comment (afedynitch) [reply src/fortran/urqmd-3.4/chromo_urqmd.f:67]: it's also cyclic as in numpy

File: src/fortran/dpmjet3.0/sources/dpmjet3.0-7.f [hunk 3]
```diff
@@ -15589,7 +15593,7 @@ SUBROUTINE DT_INITJS(MODE)
 *                = 2 DTUNUC settings                                   *
 * This version dated 16.02.96 is written by S. Roesler                 *
 *                                                                      *
-* Last change 27.12.2006 by S. Roesler.                                *
+* Last change 23.12.2023 by A. Fedynitch.                              *
```
[the remark this answers is not in the document]
Comment (afedynitch) [reply src/fortran/dpmjet3.0/sources/dpmjet3.0-7.f:15596]: fixed

File: src/fortran/dpmjet3.0/sources/phojet1.12-36c4.f [hunk 4]
```diff
@@ -34826,7 +34826,7 @@ SUBROUTINE PHO_FRAINI(IDEFAU)
       INTEGER PYCOMP
 
       IDEFAB = ABS(IDEFAU)
-
+      WRITE(6, *) 'PHO_FRAINI called with', IDEFAB
```
[the remark this answers is not in the document]
Comment (afedynitch) [reply src/fortran/dpmjet3.0/sources/phojet1.12-36c4.f:34829]: Yes

File: CMakeLists.txt [hunk 5]
```diff
@@ -420,14 +420,14 @@ f2py_add_module(_sophia
 
 
 ### dpmjet306
```
[the remark this answers is not in the document]
Comment (afedynitch) [reply CMakeLists.txt:422]: yup

File: src/fortran/dpmjet3.0/sources/dpmjet3.0-7.f [hunk 6]
```diff
@@ -15637,38 +15646,30 @@ SUBROUTINE DT_INITJS(MODE)
...
+C**anfe patch 2023.08.23
+C prevent particles from decaying
+      IF (OVwtdc) THEN
+         DO i = 1 , 35
+            IF ( i.LT.34 ) THEN
+
+               kc = PYCOMP(idxsta(i))  
+               MDCy(kc,1) = 0
+
+            ELSE IF ((ABS(idxsta(i)).EQ.3212).AND.(ISIg0.EQ.0))
```
[the remark this answers is not in the document]
Comment (afedynitch) [reply src/fortran/dpmjet3.0/sources/dpmjet3.0-7.f:15666]: ok

File: tests/test_final_state.py (hunk 1 above)
Comment (afedynitch) [reply tests/test_final_state.py:71]: done

File: examples/count_final_state_particles.ipynb
[the remark this answers is not in the document]
Comment (jncots) [reply examples/count_final_state_particles.ipynb:1]: The answer to this question: Sibyll2.3d by default leave pi0, Sigma0, and Sigma~0 as stable. Setting

```python
for pdgid in [111, 3212, -3212]:
   model.set_unstable(pdgid)
```

solves the problem. 

I checked, the problem disappears in #167, where all the particles are set stable/unstable explicitly: initially all particles set as unstable and after that only final stable particles set as stable.

[the remark this answers is not in the document]
Comment (afedynitch) [reply examples/count_final_state_particles.ipynb:1]: Thanks!! I think it is better to merge and then merge #167 instead of this handwritten list.

**Review by jncots — APPROVED** [review]

**Pull request merged**